synapse/storage/databases/main/event_federation.py

```diff
@@ -715,96 +723,238 @@ def _get_auth_chain_difference_txn(
+            in the list.
         """
 
-        def get_oldest_event_ids_with_depth_in_room_txn(
+        def get_backfill_points_in_room_txn(
             txn: LoggingTransaction, room_id: str
         ) -> List[Tuple[str, int]]:
-            # Assemble a dictionary with event_id -> depth for the oldest events
+            # Assemble a tuple lookup of event_id -> depth for the oldest events
             # we know of in the room. Backwards extremeties are the oldest
             # events we know of in the room but we only know of them because
-            # some other event referenced them by prev_event and aren't peristed
-            # in our database yet (meaning we don't know their depth
-            # specifically). So we need to look for the aproximate depth from
+            # some other event referenced them by prev_event and aren't
+            # persisted in our database yet (meaning we don't know their depth
+            # specifically). So we need to look for the approximate depth from
             # the events connected to the current backwards extremeties.
             sql = """
-                SELECT b.event_id, MAX(e.depth) FROM events as e
+                SELECT backward_extrem.event_id, event.depth FROM events AS event
                 /**
                  * Get the edge connections from the event_edges table
                  * so we can see whether this event's prev_events points
                  * to a backward extremity in the next join.
                  */
-                INNER JOIN event_edges as g
-                ON g.event_id = e.event_id
+                INNER JOIN event_edges AS edge
+                ON edge.event_id = event.event_id
                 /**
                  * We find the "oldest" events in the room by looking for
                  * events connected to backwards extremeties (oldest events
                  * in the room that we know of so far).
                  */
-                INNER JOIN event_backward_extremities as b
-                ON g.prev_event_id = b.event_id
-                WHERE b.room_id = ? AND g.is_state is ?
-                GROUP BY b.event_id
+                INNER JOIN event_backward_extremities AS backward_extrem
+                ON edge.prev_event_id = backward_extrem.event_id
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_failed_pull_attempts AS failed_backfill_attempt_info
+                ON
+                    failed_backfill_attempt_info.room_id = backward_extrem.room_id
+                    AND failed_backfill_attempt_info.event_id = backward_extrem.event_id
+                WHERE
+                    backward_extrem.room_id = ?
+                    /* We only care about non-state events because TODO: why */
```

Comment:
"state event edges" they're an old holdover and we should rip them out.

Reply:
Previous discussion thread: https://github.com/matrix-org/synapse/pull/13635#discussion_r956534576

"state event edges" doesn't elucidate anything for me? What were they before? Why did we care? I don't see any reference to them in the current code base.

Can we remove this extra condition in the SQL query?

Reply:
related: https://github.com/matrix-org/synapse/issues/12892

we used to use `event_edges` for two different sorts of "edges" (the current event DAG, but also a link to the previous state, for state events). They can be distinguished by `is_state`.

https://github.com/matrix-org/synapse/pull/12893 removes the rows with `is_state=True`, but unfortunately, because it is a background update, it's not necessarily safe to assume that it will have completed.

Reply:
*chefs kiss

Added this context to the code here 🎉

Thanks for the info @erikjohnston and @richvdh!